### ISSUE TITLE
Explore and add static checks for DAGs for early detection of common issues

### ISSUE THREAD

**Issue by omkar-foss** [apache/airflow#43176]
### Description

As per users' feedback in the Airflow Debugging Survey 2024, 48.3% of respondents chose early issue detection during execution as one of their top 2 choices.

### Use case/motivation

Goal of this issue:
- Enhance early detection of DAG issues to minimize dev time
- Some sort of static analysis similar to Ruff checks for DAGs
- Runtime analysis of DAGs if feasible can also be done to save dev time

### Related issues

Parent Issue: https://github.com/apache/airflow/issues/40975

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by omkar-foss** [comment]
In the context of this issue, DAG linting as mentioned by this article (shared by @potiuk on slack) can also be explored:

https://medium.com/@snir.isl/mastering-airflow-dag-standardization-with-pythons-ast-a-deep-dive-into-linting-at-scale-1396771a9b90

Inspired by above article, may be we can have commands like `airflow dag lint <dagname>.py` for checking common DAGs issues and `airflow dag grade <dagname>.py` for grading DAGs based on their code quality.

**Comment by potiuk** [comment]
Love that idea. We could even had some way of checking for "best practices" - like not using DB while parsing etc. This might also be then used as part of the upgrade-check mechanism that we are planning for Airflow 2-> 3 migration - see https://github.com/apache/airflow/issues/41641 cc: @Lee-W

**Comment by Lee-W** [comment]
I feel it's a bit different. 🤔 But for `not using DB while parsing`, that's something we should check in the upgrade check. But the "best practices" thing would probably be something else. Probably integrating with ruff or building our own linter would be better.

**Comment by Dev-iL** [comment]
Good to see new [airflow-specific Ruff rules](https://docs.astral.sh/ruff/rules/#airflow-air) being added (AIR3##). Nice work @Lee-W!

**Comment by Lee-W** [comment]
We could probably start discussing what the best practice is after Airflow 3.0 is released (best practice now might not reflect the latest best practice we want then)

**Comment by omkar-foss** [comment]
> We could probably start discussing what the best practice is after Airflow 3.0 is released (best practice now might not reflect the latest best practice we want then)

Yes that makes sense, let's add support for best practices linting for Airflow 3.x onwards. Just to confirm this, I took a quick opinion and got a good concensus for 3.x onwards on [slack](https://apache-airflow.slack.com/archives/C07J87PK1BK/p1739422862069399).

**Comment by Lee-W** [comment]
> [@Lee-W](https://github.com/Lee-W) would you be interested in picking up this issue? Because I suppose you have good context on Airflow static checks by adding [Ruff rules](https://github.com/apache/airflow/issues/43176#issuecomment-2571560579) for it, among other things.

Yep, implementing them is interesting, and I would definitely be interested in adding more rules haha

**Comment by Dev-iL** [comment]
Perhaps we should start a shared doc to raise ideas?

Personally, I'd love some checks related to XCOMs, and in particular in the context of task groups, to ensure they're unpacked/deserialized properly. Also, if possible, a check that custom classes returned by dags are included in the `allowed_deserialization_classes`.

**Comment by Lee-W** [comment]
we could probably just discuss here and after we have some rough idea, we can organize them to something like https://github.com/apache/airflow/issues/41641

**Comment by Dev-iL** [comment]
**Idea:**
If a `task.branch` has at least two `return`s, and exactly one of them is not `[]`, suggest that the user replaces it with `task.short_circuit`, which is simpler.

**Comment by Dev-iL** [comment]
**Idea**:
When a task/operator has a templated input that the user specifies as a string, and it contains a template to pulls the xcom of exactly one task, suggest replacing it with the taskflowy output of the task itself. For example:

```python
some_task = SomeOperator(task_id="some_task", ...)

# Before:
some_other_task = SomeOtherOperator(
    templated_field="{{ ti.xcom_pull(task_ids='some_task') }}",
)

# After:
some_other_task = SomeOtherOperator(
    templated_field=some_task.output,
)
```

**Comment by Dev-iL** [comment]
Some ideas from [this blog post](https://medium.com/apache-airflow/mastering-airflow-dag-standardization-with-pythons-ast-a-deep-dive-into-linting-at-scale-1396771a9b90):
- The DAG must not have top-level calls to expensive classes/services.
- The DAG object should be instantiated using a context manager.
- The DAG file name should match its `dag_id`.

**Comment by Dev-iL** [comment]
@zach-overflow 
>     * All `dag_id`'s should be unique (AKA prevent name collisions and detect them at validation time).
I'm not sure Ruff supports multi-file checks, so I suspect this might only be applicable to situations where multiple DAGs are defined in a single script. IIRC this can be detected by a [DagBag unit test](https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#unit-tests).

>     * Lastly -- and this might be a larger scope suggestion -- it would be _really_ useful if the Airflow linter could validate the user's DAG code against any user-defined task policy or dag policy (as defined in their [cluster policy](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/cluster-policies.html#cluster-policies)).
I think this too belongs in the realm of  checks to be added to `DagBag` and/or unit tests. 

>     * Optionally check whether DAG definitions are serializable or not
Could you please elaborate (mainly out of curiosity):
- What common pattern might make a DAG definition non-serializable?
- In what contexts does this matter?
- Is this supposed to be a static or a DagBag-related test? If static - how can this be done without trying to serialize the DAG and seeing if it works?

--------

All in all I these are some good ideas to be included in `DagBag`, some variation of DagBag that's used for DAG validation, or at the very least - unit test examples to be added to the docs.

**Comment by snirisl** [comment]
> Some ideas from [this blog post](https://medium.com/apache-airflow/mastering-airflow-dag-standardization-with-pythons-ast-a-deep-dive-into-linting-at-scale-1396771a9b90):
> 
> * The DAG must not have top-level calls to expensive classes/services.
> * The DAG object should be instantiated using a context manager.
> * The DAG file name should match its `dag_id`.

Thanks @Dev-iL for referencing my blog post 🙏 Another thing we can add is consider implementing a rule, adding to the third rule above, that dag id should be all lower case characters. Since DAG IDs should match the file names, enforcing this one is making sure DAG files to be compatible with the Python convention for file names.

**Comment by Dev-iL** [comment]
- Check that paths to temporary files/folders created in a task aren't returned (in XCOM) - since it cannot be guaranteed that downstream tasks will run on the same worker, and therefore will have access to the temporary path.

**Comment by Lee-W** [comment]
Hey, sorry for taking so long. I listed down all the rules above and my thoughts.

---


#### 1. Use `task.short_circuit` when `task.branch` has multiple `return`s and exactly one of them is not `[]`

* Should we do it?
    * Is it a typical case?
* Is it doable?
    * Should be

<https://github.com/apache/airflow/issues/43176#issuecomment-2667604469>

#### 2. If `SomeOtherOperator(some_template_field="{{ ti.xcom_pull(task_ids=['task1']) }}")` is used, replace it with `...... some_template_field=task1.output` instead

* Should we do it?
    * Sounds good
* Is it doable?
    * Possibly
    * Need to define the exact list of template_field

<https://github.com/apache/airflow/issues/43176#issuecomment-2667826944>

#### 3. The Dag must not have top-level calls to expensive classes/services.

* Should we do it?
    * Sounds good
* Is it doable?
    * Possibly
    * More detail needed. What are "expensive classes/services"? We need an exact list to check, and it needs to be generalized enough

> Ensure function-level imports are used in DAG files when possible (again, an example of this kind of rule may be found in the airflint repo here)



<https://github.com/apache/airflow/issues/43176#issuecomment-2667878168>
<https://github.com/apache/airflow/issues/43176#issuecomment-2669919039>

#### 4. The DAG object should be instantiated using a context manager.

* Should we do it?
    * Why? taskflow also works fine?
* Is it doable?
    * Should be

<https://github.com/apache/airflow/issues/43176#issuecomment-2667878168>

#### 5. The Dag file name should match its dag_id.

* Is it doable?
    * I doubt it's doable

<https://github.com/apache/airflow/issues/43176#issuecomment-2667878168>

#### 6. All `dag_id` s should be unique

* Is it doable?
    * Single file. Possible. But will need extra effort to dig into it
    * Cross files. Probably not.

<https://github.com/apache/airflow/issues/43176#issuecomment-2669919039>

#### 7. Require Variables passed to operator constructors are only passed as Jinja templates (per the official best practice section here).

* Should we do it?
    * Sounds good
* Is it doable?
    * Should be

<https://github.com/apache/airflow/issues/43176#issuecomment-2669919039>

#### 8. it would be really useful if the Airflow linter could validate the user's DAG code against any user-defined task policy or dag policy

* Is it doable?
    * I doubt it's doable

<https://github.com/apache/airflow/issues/43176#issuecomment-2669919039>

#### 9.  Optionally check whether DAG definitions are serializable or not

* Is it doable?
    * I doubt it's doable

<https://github.com/apache/airflow/issues/43176#issuecomment-2672255689>

#### 10. tmp path not returned by tasks

* Should we do it?
    * not sure whether it's a common case
* Is it doable?
    * Should be

<https://github.com/apache/airflow/issues/43176#issuecomment-2695069193>

#### 11. Don't allow implicit python task decorators, `@task`, and instead require specifying `@task.python`.

* Should we do it?
    * netural
* Is it doable?
    * Should be

<https://github.com/apache/airflow/issues/43176#issuecomment-2737216835>

**Comment by Dev-iL** [comment]
> #### 1. Use `task.short_circuit` when `task.branch` has multiple `return`s and exactly one of them is not `[]`
>   * Is it a typical case?

I ran into this myself, no idea how common it is but I bet I'm not the only one. Can be a low priority check.

> #### 2. If `SomeOtherOperator(some_template_field="{{ ti.xcom_pull(task_ids=['task1']) }}")` is used, replace it with `...... some_template_field=task1.output` instead
>   * Need to define the exact list of template_field

Did you ask for a list to make this more general? What I had in mind applies only to the output of a task, when the xcom key is not provided (and thus set to `return_value`) - in other words this check should detect templates of the form

```
{{ <task instance object>.xcom_pull(task_ids=<iterable of length 1>[, key="return_value"])` }}
```

> #### 3. The Dag must not have top-level calls to expensive classes/services.
>   * More detail needed. What are "expensive classes/services"? We need an exact list to check, and it needs to be generalized enough

Expensive = takes either absolutely or relatively long to load. The main example that comes to mind is pandas. However, I think this is something that's more reasonable to live inside a unit test (like the dagbag one), or a threshold used by the processor to emit warnings. Re generalization - it's simpler to define what *is* allowed (e.g. airflow or standard library imports) rather than what isn't - but this approach will flag much of the existing code.

> #### 4. The DAG object should be instantiated using a context manager.
>   * Why? taskflow also works fine?

Taskflow is fine and context manager is fine too. But perhaps `dag = DAG(....)` should be discouraged.

> #### 5. The Dag file name should match its dag_id.
> #### 6. All `dag_id` s should be unique

Can be checked by a unit test (dagbag).

> #### 8. it would be really useful if the Airflow linter could validate the user's DAG code against any user-defined task policy or dag policy

Potentially possible via a dagbag test, assuming it processes dag policies (might require dev effort to support).

> #### 10. tmp path not returned by tasks 
>   * not sure whether it's a common case

This is something users might run into when scaling up from a single-worker setup - where all of a sudden DAGs start failing. If we can (add this easily I say we should.

> #### 11. Don't allow implicit python task decorators, `@task`, and instead require specifying `@task.python`.
>   * netural

Like every other rule, it could be opt-in.

**Comment by potiuk** [comment]
Yep. Good start!

**Comment by Lee-W** [comment]
> Did you ask for a list to make this more general? What I had in mind applies only to the output of a task, when the xcom key is not provided (and thus set to return_value) - in other words this check should detect templates of the form

I'm asking for a list of fields we need to check to make it more specific. Not all string in `{{ ... }}` format should be checked. We need a concrete idea on what to be checked.

> Expensive = takes either absolutely or relatively long to load. The main example that comes to mind is pandas. However, I think this is something that's more reasonable to live inside a unit test (like the dagbag one), or a threshold used by the processor to emit warnings. Re generalization - it's simpler to define what is allowed (e.g. airflow or standard library imports) rather than what isn't - but this approach will flag much of the existing code.

Yep, but we'll have to define a concrete list. Maybe worth create another issue to get this list. And if `airflow....DAG` or `airflow...dag` is imported and these kind of operations exists, we raise a warning. But we need a concrete list of "these kinds of operation.

> Taskflow is fine and context manager is fine too. But perhaps dag = DAG(....) should be discouraged.

Totally agree, this one shouldn't be hard.

> Can be checked by a unit test (dagbag).

Yep, then this would be out of Ruff's capability.

> Potentially possible via a dagbag test, assuming it processes dag policies (might require dev effort to support).

Same as above

**Comment by wjddn279** [comment]
additional idea for lint rule

Prevents values that can change at runtime from being used as arguments for Airflow DAGs or tasks. 

ref: https://github.com/apache/airflow/pull/59430 

While this PR performs runtime checks, this would be an excellent case for lint-time checking as well.

**Comment by Lee-W** [comment]
Yep, if we have a list of things like this, it's more doable on the ruff side. Thanks!

**Comment by Dev-iL** [comment]
Several updates on this:

- (User-facing) Ruff rules are coming along nicely. Astral seem to be fine with us adding rules as long as we don't overwhelm them (a good rule of thumb is one rule at a time) and follow the Airflow [community guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/24_proposing_best_practices_and_air_rules.rst) (mainly - achieving consensus on the devlist before submitting the implementation to Ruff).
- Implementing new Ruff rules for Airflow is almost effortless using the [SKILL made for this purpose](https://gist.github.com/Dev-iL/3ee77257fff578b884a0f7492d082093); the limiting factor at the moment is a shortage of inspection ideas.

It is worth mentioning we also have several [developer-facing AST rules available in the Airflow repo itself](https://github.com/apache/airflow/blob/main/scripts/ci/prek/check_migration_patterns.py), so if anyone is looking to add non-Ruff inspections, feel free to explore this technique.

I think with the exception of (tooling for) the "Runtime analysis of DAGs if feasible" bullet, this open-ended issue seems mostly resolved to me.

**Comment by nredd** [comment]
Thanks for the update @Dev-iL!! Excited to start iterating 😃
